### PR TITLE
CZI: remaining fixes for ticket 12935

### PR DESCRIPTION
**Pull request by melissalinkert**
See https://trac.openmicroscopy.org/ome/ticket/12935#comment:8 and https://trello.com/c/y6KpD3sU/6-czi-tickets

325a4c9 addresses point 2 on the ticket comment, d026924 addresses point 3, 8a596d1 addresses point 4, and 0242c1d addresses point 5.

To test points 2 and 3, compare ```showinf -nopix -omexml``` against the ```Info``` tab in ZEN for each of the following files.  The ```Stage Position``` in ZEN should match ```PositionX``` and ```PositionY``` in the OME-XML, and the ```Detector Gain``` for each channel in ZEN should match the ```Gain``` for each channel's linked ```Detector``` in the OME-XML.  File list:

```
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/Fluo#1_ZEN2-1-Airyscan.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/Lambda1.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/LambdaImage.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/LambdaImage_Linear unmixing.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/LambdaImage_Spectral channel extraction2CHannel.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/LAmbda+Rand+Tpmt.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/MT-Tilescan2-3stitched.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/MT-Tilescan2-3unstitched.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/MT-ZStack5_Position4_Time3.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/OnlineFingerprinting_2Channel.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/SIM_nicht verrechnet.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/SIM_verrechnet.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/Tile4-4Zstack5_Gestitchen.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/TileZstack5_Nicht Gestitchen.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/Time_Bleaching_Marker_BleachROI.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/Timeserie_GraphikElemente.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/WF.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/5Positionen.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/channel_3.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/HDR_Illu.czi
zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015/HDRsampling.czi
```

To test point 4, verify that the ```test-automated``` repository test on ```curated/zeiss-czi/zeiss/ZEN 2.1 black LSM 880 -May 2015``` fails on ```testValidXML``` for ```SIM_nicht verrechnet.czi``` without this change, and passes with it.  It may be worth a deeper discussion of how to handle/document modulo annotation ```Label``` vs ROI ```Label``` at some point, but this solves the immediate issue of not being able to test files that have a modulo annotation ```Label```.

To test point 5, check the output of ```showinf -nopix -omexml``` for each of the following files.  Without this change, there should be XML validation errors indicating that there are ```Detector```s with duplicate ```ID```s.  With this change, there should be no errors and no duplicate ```ID```s.  File list:

```
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/LambdaCh10T3/LambdaCh10T3.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/LambdaCh10T3ConvertedForZENblack/LambdaCh10T3ConvertedForZENblack.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/LambdaCh10T1/LambdaCh10T1.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/LambdaCh10T1ConvertedForZENblack/LambdaCh10T1ConvertedForZENblack.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/AiryscanCh3T1Z1/AiryscanCh3T1Z1.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/AiryscanCh3T1Z5/AiryscanCh3T1Z5.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/AiryscanCh3T3Z1/AiryscanCh3T3Z1.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/AiryscanCh3T3Z5/AiryscanCh3T3Z5.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/AiryscanProcessedCh3T1Z1/AiryscanProcessedCh3T1Z1.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/AiryscanProcessedCh3T1Z5/AiryscanProcessedCh3T1Z5.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/AiryscanProcessedCh3T3Z1/AiryscanProcessedCh3T3Z1.czi
zeiss-czi/zeiss/ZEN 2.1 blue LSM 800 -May2015/AiryscanProcessedCh3T3Z5/AiryscanProcessedCh3T3Z5.czi

### CODE REVIEW

**Comment by melissalinkert** [comment]
As discussed with @sbesson, we do have sample OME-TIFFs that use ```Label``` on ```Modulo```, namely ```SPIM-ModuloAlongZ.ome.tiff```.  In that case, the ```Label``` has been created in such a way that it will validate as a ```Shape``` (note discrepancy with ```Label``` definitions in http://www.openmicroscopy.org/site/support/ome-model/developers/6d-7d-and-8d-storage.html):

```
<ModuloAlongZ Type="angle" Unit="degree">
                  <Label ID="Shape:0" Text="0" X="0" Y="0"/>
                  <Label ID="Shape:1" Text="90" X="90" Y="90"/>
                  <Label ID="Shape:2" Text="180" X="180" Y="180"/>
                  <Label ID="Shape:3" Text="270" X="270" Y="270"/>
</ModuloAlongZ>
```

which is why ```testValidXML``` did not fail on the OME-TIFF samples.  I'd be open to reverting 8a596d1 and instead updating the ```Modulo``` annotation creation to create ```Label```s similar to those in the sample files, if that is more agreeable.  Also happy to revert and open a card for 5.6.0 if this change needs more discussion.

File: components/formats-api/src/loci/formats/services/OMEXMLServiceImpl.java [hunk 1]
```diff
@@ -554,6 +554,57 @@ public boolean validateOMEXML(String xml, boolean pixelsHack) {
...
+          nodes = annotationRoot.getElementsByTagName("ModuloAlongT");
+          if (nodes.getLength() > 0) {
+            ((XMLAnnotation) annotation).setValue("");
+          }
+        }
+      }
+      omexml.setRoot(root);
+      xml = getOMEXML(omexml);
+    }
+    catch (ServiceException e) {
```
Comment (mtbc) [inline components/formats-api/src/loci/formats/services/OMEXMLServiceImpl.java]: just use multi-catch? (hmmm, can't find where we say which version of Java we require...)

File: components/formats-gpl/src/loci/formats/in/ZeissCZIReader.java [hunk 2]
```diff
@@ -1826,6 +1829,28 @@ private void translateInformation(Element root) throws FormatException {
...
+          for (int p=0; p<positions.getLength(); p++) {
+            Element position = (Element) positions.item(p);
+            String x = position.getAttribute("X");
+            String y = position.getAttribute("Y");
+            String z = position.getAttribute("Z");
+            if (nextPosition < positionsX.length && positionsX[nextPosition] == null) {
+              positionsX[nextPosition] = new Length(DataTools.parseDouble(x), UNITS.MICROM);
+              positionsY[nextPosition] = new Length(DataTools.parseDouble(y), UNITS.MICROM);
+              positionsZ[nextPosition] = new Length(DataTools.parseDouble(z), UNITS.MICROM);
+              nextPosition++;
```
Comment (dgault) [inline components/formats-gpl/src/loci/formats/in/ZeissCZIReader.java]: There are 2 failures for the units in https://ci.openmicroscopy.org/view/Failing/job/BIOFORMATS-DEV-merge-full-repository/577

PlanePositions: FAILED (X position unit series 0, plane 0 (expected reference frame, actual µm)) for 
curated/zeiss-czi/nottingham/incorrect-colors/SpinalChord63x_1 SIMcrop_Structured Illumination.czi and curated/zeiss-czi/nottingham/incorrect-colors/Image 6_Structured Illumination.czi

Looking at the files in Zen Im seeing a stage position of X: -3.28 mm and Y: -2.63 mm for SpinalChord and X: 13.45 mm Y: -3.52 mm for Image6

Checking some of the other stage values from Zen against the new config PR and the updated configs look to be correct. Did these 2 files simply miss the config update for some reason?

Comment (melissalinkert) [inline components/formats-gpl/src/loci/formats/in/ZeissCZIReader.java]: Sorry, yes, those two files were missed by accident.  Configurations for both should be fixed with https://github.com/openmicroscopy/data_repo_config/pull/220/commits/8bde52c329cab924098483463b20c1abe0793f53

File: components/formats-api/src/loci/formats/services/OMEXMLServiceImpl.java (hunk 1 above)
Comment (melissalinkert) [inline components/formats-api/src/loci/formats/services/OMEXMLServiceImpl.java]: Commit pushed; closest minimum version documentation we have I think is http://blog.openmicroscopy.org/tech-issues/future-plans/2015/10/15/java-6/

**Comment by dgault** [comment]
With the latest commits and config updates the builds and tests are all green. The code changes look good and make sense based on the Trac issues.

Using the recommended sample files the stage positions and gain now match those displayed in the info window in Zen. The XML validation issues for Label and duplicate ID's are also no longer present.

It seems like it might be worth some follow up discussion on the Modulo labels for 5.6 but for the purposes of this PR the immediate issue is addressed. PR looks good to merge.

**Comment by sebi06** [comment]
Hi guys, I came across that due to reading one of your trello cards: https://trello.com/c/x5ihZ1WS that had to do with XY-Stage position of CZI files and the used units.
Please let me know when you need more info here. And I saw that a lot of the CZI files are a bit older or CZI's written by ZEN Black, which is not 100% the official CZI ... (i know this sound confusing, but that the way it is)
So I am always concerned that a fix based on those ZEN Black CZIs will break stuff for the official CZI file format.

**Comment by imagejan** [comment]
@sebi06 wrote:
> CZI's written by ZEN Black, which is not 100% the official CZI 

😆 Zeiss not sticking to their own specification? You're kidding!

But these files are out there, and many people have data saved with ZEN Black, so I think it's good to keep supporting them, no?

**Comment by sebi06** [comment]
No, that is not the case. But it is a matter of fact that a CZI from Zen
Black is not totally identical to one written from ZEN Blue yet
Therefore every fix inside BioFormats based on a “Black“ CZI has to be test
against the latest and official CZI from Zen Blue as well.
All of this has historical reasons and has no consequences in 99% of what
people do.
Therefore, no panic. Just keep that in mind in the unlikely case something
behaves oddly.

Sebi (from Zeiss)


Am 09.08.2017 3:05 nachm. schrieb "Jan Eglinger" <notifications@github.com>:

**Pull request merged**